python/pyspark/sql/pandas/types.py

```diff
@@ -74,6 +74,8 @@ def to_arrow_type(dt):
         arrow_type = pa.struct(fields)
     elif type(dt) == NullType:
         arrow_type = pa.null()
+    elif isinstance(dt, UserDefinedType):
+        arrow_type = to_arrow_type(dt.sqlType())
```

Comment:
Seems like so far only ArrayType is supported with UTD. Is it possible to block other complex data types with UTD, like struct or map including UDT or array of array of UDT, here? Otherwise, I guess we will see a weird error from PyArrow.

Reply:
Also, this is used in `SparkConversionMixin.createDataFrame` and `PandasConversionMixin.toPandas` too, could you check how this change affects them?

Reply:
Additional info:

SparkConversionMixin.createDataFrame -> to_arrow_type
PandasConversionMixin.toPandas -> to_arrow_schema -> to_arrow_type

we will add unit tests on it. (P.S. I'm co-working to eddyxu.)

Reply:
This PR is too big. I plan to separate this part into a subtask:

https://issues.apache.org/jira/browse/SPARK-34771

Reply:
I've checked `SparkConversionMixin.createDataFrame` and `PandasConversionMixin.toPandas`.

Both of them do not actually support arrow. When `spark.sql.execution.arrow.enabled` is `true`,  it will fallback to mode without arrow optimization.

Reply:
I will create subsequent PR to solve [SPARK-34771](https://issues.apache.org/jira/browse/SPARK-34771) if possible.

Reply:
PR created: https://github.com/apache/spark/pull/32026/files